lib/compress/zstd_fast.c

```diff
@@ -490,26 +535,33 @@ size_t ZSTD_compressBlock_fast_dictMatchState_generic(
                 ip0++;
                 ZSTD_storeSeq(seqStore, (size_t) (ip0 - anchor), anchor, iend, REPCODE1_TO_OFFBASE, mLength);
                 break;
-            } else if (matchIndex <= prefixStartIndex) {
-                /* We only look for a dict match if the normal matchIndex is invalid */
+            }
+
+            if (dictTagsMatch) {
+                /* Found a possible dict match */
+                const U32 dictMatchIndex = dictMatchIndexAndTag >> ZSTD_SHORT_CACHE_TAG_BITS;
                 const BYTE* dictMatch = dictBase + dictMatchIndex;
                 if (dictMatchIndex > dictStartIndex &&
                     MEM_read32(dictMatch) == MEM_read32(ip0)) {
-                    /* found a dict match */
-                    U32 const offset = (U32) (curr - dictMatchIndex - dictIndexDelta);
-                    mLength = ZSTD_count_2segments(ip0 + 4, dictMatch + 4, iend, dictEnd, prefixStart) + 4;
-                    while (((ip0 > anchor) & (dictMatch > dictStart))
-                           && (ip0[-1] == dictMatch[-1])) {
-                        ip0--;
-                        dictMatch--;
-                        mLength++;
-                    } /* catch up */
-                    offset_2 = offset_1;
-                    offset_1 = offset;
-                    ZSTD_storeSeq(seqStore, (size_t) (ip0 - anchor), anchor, iend, OFFSET_TO_OFFBASE(offset), mLength);
-                    break;
+                    /* To replicate extDict parse behavior, we only use dict matches when the normal matchIndex is invalid */
+                    if (matchIndex <= prefixStartIndex) {
```

Comment:
It seems weird to put this condition after checking that we have a match.

Maybe the compiler generates slightly better code in this case for some dumb reason, but it seems like clearly the wrong order.

Could you do `dictTagsMatch && matchIndex <= prefixStartIndex` on line 540?

Reply:
It's really unfortunate that clang doesn't like `dictTagsMatch && matchIndex <= prefixStartIndex` (or the reverse). I tried one permutation [here](https://github.com/facebook/zstd/pull/3152/commits/a243a5343dfdbb14ed08a216e2661429438aeaf2) (tried the other permutations locally) and it turned out to be a regression for some clang scenarios.

I think the problem is that if you combine the branches like that, you have to add a branch for the ctx search: 
```
if (dictTagsMatch && matchIndex <= prefixStartIndex) {
    // check dict match
} else if (matchIndex > prefixStartIndex) {
    // check ctx match
}
```
The extra branch appears to mess up codegen.

The other option is to split it into two nested branches:
```
if (matchIndex <= prefixStartIndex) {
    if (dictTagsMatch) { 
        // check dict match
    }
} else {
    // check ctx match
}
```
Either clang or gcc didn't like that (don't remember which, it was one of about a dozen setups I tested locally without pushing to github).

I can definitely add a comment explaining the situation though, agree it's unclear.

Reply:
In a separate PR (not anytime soon, but someday) it would be interesting to test whether we can just remove the `matchIndex <= prefixStartIndex` check entirely. There's no fundamental reason why we have to parse the same as extDict.

Reply:
One more point of clarification: we can set up the branches any way we want to get large cold dict wins. The hyper-tuning is just about preventing hot dict regressions, because we really want to avoid gating short cache if possible.

Reply:
Man I dunno this seems janky to me. You're adding a branch and it's weird that the dict search is coming ahead of the working ctx search. What about:

```
if (matchIndex >= prefixStartIndex) {
  if (MEM_read32(match) == MEM_read32(ip0)) {
    // found ctx match
  }
} else if (dictTagsMatch) {
  if (MEM_read32(dictMatch) == MEM_read32(ip0)) {
    // found dict match
  }
}

```

Reply:
I think I measured this and it regressed some scenario but I don't remember for sure. Will measure and report back, would be great if we can clean this up.

Reply:
Here are the numbers. The colors aren't consistent between graphs, be careful to look at each legend (sorry about that, not sure why plotly didn't order things the same way for each one).

It looks like your version is a few % win for clang and a few % loss for gcc. In the compiler=gcc and DictSize=4K scenario, your version takes the win to 0 or slightly negative. However, these numbers are *on top* of the already-landed fast DMS pipeline, so it's ok if some scenarios are slightly negative.

![newplot (10)](https://user-images.githubusercontent.com/12179121/174331903-87b1a226-83d1-4f54-8345-aa349cdd3fbf.png)
![newplot (11)](https://user-images.githubusercontent.com/12179121/174331905-3f44b554-529e-4a7e-be79-83dea9d85cb4.png)
![newplot (12)](https://user-images.githubusercontent.com/12179121/174331906-83b0d947-ca2b-47b1-8d6f-7379f3b8e506.png)

Reply:
Discussed offline: the proposed change *shouldn't* affect perf... but of course in the real world we know our compilers can be a little picky sometimes :) @felixhandte and I decided to land as-is for now, and at some point we'll see if we can figure out why GCC doesn't like the less janky version.